File: l2-pa-project-fortaCB-cac63224d1987995bf3ed5804e86e499def30513-cac63224d1987995bf3ed5804e86e499def30513/l2-pa-project-fortaCB-cac63224d1987995bf3ed5804e86e499def30513-cac63224d1987995bf3ed5804e86e499def30513/main.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include "def.h"
using namespace std;

#define START_FEN  "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

bool isMove(string s) {
    if (s.length() > 5)
        return false;
    if (s[0] < 'a' || s[0] > 'h')
        return false;
    if (s[1] < '1' || s[1] > '8')
        return false;
    if (s[2] < 'a' || s[2] > 'h')
        return false;
    return !(s[3] < '1' || s[3] > '8');
}

int parseTime(string line) {
    stringstream sin(line);
    int time = -1;
    string s;
    sin >> s >> time;
    return time;
}

Move parseMove(string s, Board *b);

void writeMove(Board *pos,int time){
	Move m;
	if(pos->ply == 0 && pos->side==WHITE){
			cout<<"move e2e4"<<endl;
			makeMove(parseMove("e2e4",pos),pos);
			return;
	}
	if(pos->ply == 1 && pos->side==BLACK){
			cout<<"move e7e5"<<endl;
			makeMove(parseMove("e7e5",pos),pos);
			return;
	}
	int depth=4;
	int i,sum=0;
	for(i=wP;i<=bK;i++){
		sum+=pos->pcsNum[i];
	}
	if(time>12000 && pos->ply>30)
		depth=5;
	if(sum<20)
		depth=5;
	if(sum<10)
		depth=6;
	if(time<3000)
		depth--;

	m = alphaBetaNegamax(-1e9,1e9, depth, pos, *NOMOVE);
    if (m.compareTo(NOMOVE) || m.score ==- 10000) {
            cout << "resign" <<endl;
    } else {
            makeMove(m, pos);
            cout<<"move "<<prMove(m)<<endl;
    }
}

int main() {
    Board *b = new Board;
    allInit();
    parseFen(START_FEN, b);
    int mode, time, otime;
    string line;
    Move m;

    while (true) {
        getline(cin, line);
        if (line.find("post") != string::npos)
            continue;
        if (line.find("hard") != string::npos)
            continue;
        if (line.find("easy") != string::npos)
            continue;
        if (line.find("undo") != string::npos)
            continue;
        if (line.find("time") != string::npos) {
            time = parseTime(line);
            continue;
        }
        if (line.find("otim") != string::npos) {
            otime = parseTime(line);
            continue;
        }

        // dezactivam modul care trimite SIGINT SIGTERM SAN
        if (line.find("protover") != string :: npos) {
            cout << "feature sigint=0" << endl;
            cout << "feature san=0" << endl;
            cout << "feature sigterm=0" << endl;
        }

        if (line.find("new") != string :: npos) {
            parseFen(START_FEN, b);
            mode = 1;  // black
            continue;
        }

        // setam modurile
        if (line.find("black") != string :: npos) {
            mode = 1;  // black
            getline(cin, line);
        }
        if (line.find("white") != string :: npos) {
            mode = 0; // white
            getline(cin, line);
        }

        if (line.find("go") != string :: npos) {
            writeMove(b,time);
            continue;
        }

        if (isMove(line)) {
            Move move = parseMove(line, b);
            makeMove(move, b);
           	writeMove(b,time);
            continue;
        }

        if (line.find("force") != string :: npos) { // daca intram in modul force engine ul nu mai face mutari singur
            while (true) {
                getline(cin, line);

                if (line.find("black") != string :: npos) { // asteptam "go" sau "new" ca sa iesim din
                    // force mod
                    mode = 1;  //black
                    break;
                }
                if (line.find("white") != string :: npos) { // asteptam "go" sau "new" ca sa iesim din
                    // force mod
                    mode = 0;  // white
                    break;
                }
                if (line.find("go") != string :: npos){
                    m = alphaBetaNegamax(-1e9,1e9, 4, b, *NOMOVE);
                    writeMove(b,time);
                    break;
                }
                if (line.find("new") != string :: npos) {
                    parseFen(START_FEN, b);
                    mode = 1;  // black
                    break;
                }

                // ignoram comenzile care nu ne intereseaza
                if (line.find("post") != string :: npos)
                    continue;
                if (line.find("hard") != string :: npos)
                    continue;
                if (line.find("easy") != string :: npos)
                    continue;
                if (isMove(line)) {
                    Move move = parseMove(line, b);
                    makeMove(move, b);
                    continue;
                }
            }
        }

        if (line.find("quit") != string :: npos)
            break;
    }
}

Move parseMove(string s, Board *b) {
    if (!isMove(s))
        return *NOMOVE;

    int fileFrom = s[0] - 'a';
    int rankFrom = s[1] - '1';
    int fileTo = s[2] - 'a';
    int rankTo = s[3] - '1';
    int from = FR2SQ(fileFrom, rankFrom);
    int to = FR2SQ(fileTo, rankTo);
    int captured = b->pieces[to]; // daca e ceva ok bine, daca nu va fi EMPTY adica 0

    int promoted = EMPTY;
    if (s.size() == 5) {
        switch (s[4]) {
            case 'q': promoted = b->side == WHITE ? wQ : bQ;
                break;
            case 'r': promoted = b->side == WHITE ? wR : bR;
                break;
            case 'b': promoted = b->side == WHITE ? wB : bB;
                break;
            case 'k': promoted = b->side == WHITE ? wK : bK;
                break;
            default: return *NOMOVE;
        }
    }

    bool castle = false;
    if (b->side == BLACK){

        if (b->castlePerm & BKCA)
            if (b->pieces[F8] == EMPTY && b->pieces[G8] == EMPTY)
                if (!sqAttacked(E8, WHITE, b) && !sqAttacked(F8, WHITE, b))
                    castle = true;
        if (b->castlePerm & BQCA)
            if (b->pieces[D8] == EMPTY && b->pieces[C8] == EMPTY && b->pieces[B8] == EMPTY)
                if (!sqAttacked(E8, WHITE, b) && !sqAttacked(D8, WHITE, b))
                    castle = true;
    } else {
        if (b->castlePerm & WKCA)
            if (b->pieces[F1] == EMPTY && b->pieces[G1] == EMPTY)
                if (!sqAttacked(E1, BLACK, b) && !sqAttacked(F1, BLACK, b))
                    castle = true;
        if (b->castlePerm & WQCA)
            if (b->pieces[D1] == EMPTY && b->pieces[C1] == EMPTY && b->pieces[B1] == EMPTY)
                if (!sqAttacked(E1, BLACK, b) && !sqAttacked(D1, BLACK, b))
                    castle = true;
    }

    bool pawnStart = false;
    if (isPawn[b->pieces[from]]) {
        if (b->side == WHITE && ranks[from] == RANK_2 && ranks[to] == RANK_4) {
            pawnStart = true;
        }
        if (b->side == BLACK && ranks[from] == RANK_7 && ranks[to] == RANK_5) {;
            pawnStart = true;
        }
    }

    bool enPas = false;
    if(b->enPas == to)
        enPas = true;

    Move *m = new Move(from, to, captured, promoted, castle, pawnStart, enPas);
    return *m;
}

File: l2-pa-project-fortaCB-cac63224d1987995bf3ed5804e86e499def30513-cac63224d1987995bf3ed5804e86e499def30513/l2-pa-project-fortaCB-cac63224d1987995bf3ed5804e86e499def30513-cac63224d1987995bf3ed5804e86e499def30513/Main.java
import java.util.*;

public class Main {
	// verifica daca s-a primit o mutare valida ca input
	public static int isMove(String s, Board b) {
		if (s.length() > 5)
			return 0;
		if (s.charAt(0) < 'a' || s.charAt(0) > 'h')
			return 0;
		if (s.charAt(1) < '1' || s.charAt(1) > '8')
			return 0;
		if (s.charAt(2) < 'a' || s.charAt(2) > 'h')
			return 0;
		if (s.charAt(3) < '1' || s.charAt(3) > '8')
			return 0;
		return 1;
	}
	/**
	 * Verifica daca exista mutari realizabile de pioni
	 * @param b variabila de tip Board necesara pentru accesarea tablei si functiilor
	 * @param mode culoarea
	 * @return 0 daca nu exista mutari valabile, 1 daca exista
	 */
	//Nu este folosita in etapa aceasta a proiectului
	public static int makeAPMove(Board b, String mode) {
		String s = b.pawnMove(mode);
		// daca functia a intors "no moves" nu mai sunt mutari valide pentru pioni 
		if (s.contains("no")) 
			return 0;
		else {
			System.out.println("move " + s);
			b.move(s);
		}
		return 1;
	}

	// Verifica daca se poate face o mutare valida a nebunului
	// Similar cu makeAPMove ca parametri si return
	//Nu este folosita in etapa aceasta a proiectului
	public static int makeABMove(Board b, String mode) {
		String s = b.bishopMove(mode);
		if (s.contains("no"))
			return 0;
		else {
			System.out.println("move " + s);
			b.move(s);
		}
		return 1;
	}

	// verifica daca regina poate face miscari valide pe tabla
	// Similar cu makeAPMove ca parametri si return
	//Nu este folosita in etapa aceasta a proiectului
	public static int makeAQMove(Board b, String mode) {
		String s = b.queenMove(mode);
		if (s.contains("no"))
			return 0;
		else {
			System.out.println("move " + s);
			b.move(s);
		}
		return 1;
	}

	// verifica daca se poate face o miscare valida pe tabla cu tura
	// Similar cu makeAPMove ca parametri si return
	//Nu este folosita in etapa aceasta a proiectului
	public static int makeARMove(Board b, String mode) {
		String s = b.rookMove(mode);
		if (s.contains("no"))
			return 0;
		else {
			System.out.println("move " + s);
			b.move(s);
		}
		return 1;
	}

	// Verifica daca se poate face o mutare valida cu calul
	// Similar cu makeAPMove ca parametri si return
	//Nu este folosita in etapa aceasta a proiectului
	public static int makeAKMove(Board b, String mode) {
		String s = b.knightMove(mode);
		if (s.contains("no")) 
			System.out.println("resign");
		else {
			System.out.println("move " + s);
			b.move(s);
		}
		return 1;
	}
	// Verifica daca se poate face o mutare valida cu regele
	// Similar cu makeAPMove ca parametri si return
	public static int makeAKingMove(Board b, String mode) {
		String s = b.kingMove(mode);
		if (s.contains("no"))
			return 0;
		else {
			System.out.println("move " + s);
			b.move(s);
		}
		return 1;
	}
	// Verifica miscarile tuturor tipurilor de piese pe rand
	// Se introduc intr-o lista toate miscarile de la fiecare piese si se apeleaza metoda makeMove din clasa Board
	// Aceasta metoda intoarce cea mai buna mutare dintre cele introduse in lista
	public static int makeAMove(Board b, String mode) {
		String s;
		// Se creeaza o lista cu cele mai bune miscari de la fiecare piesa
		ArrayList<String> l = new ArrayList<>();
		// Se verifica miscarile pionilor
		s = b.pawnMove(mode);
		if (!s.contains("no"))
			l.add(s);
		//Se verifica miscarile regelui
		s = b.kingMove(mode);
		if (!s.contains("no"))
			l.add(s);
		//Se verifica miscarile reginei
		s = b.queenMove(mode);
		if (!s.contains("no"))
			l.add(s);
		//Se verifica miscarile turei
		s = b.rookMove(mode);
		if (!s.contains("no"))
			l.add(s);
		//Se verifica miscarile nebunului
		s = b.bishopMove(mode);
		if (!s.contains("no"))
			l.add(s);
		//Se verifica miscarile calului
		s = b.knightMove(mode);
		if (!s.contains("no"))
			l.add(s);
		//Se aplica functia makeMove pentru a intoarce cea mai buna mutare dintre toate
		s = b.makeMove(l, mode);
		//Daca nici o piesa nu poate realiza o mutare se da resign 
		if (s.contains("no")) {
			System.out.println("resign");
			return 0;
		}
		//Se afiseaza mutarea si se realizeaza mutarea in matrice folosing functia move
		System.out.println("move " + s);
		b.move(s);
		return 1;
	}

	public static void main(String[] args) throws InterruptedException {

		// TODO Auto-generated method stub
		// Se citesc comenzile de la tastatura. Xboard-ul da comenzi la Stdin 
		Scanner sc = new Scanner(System.in);
		// Culoarea este stabilita automat ca negru.
		String line,  mode = "black";
		// Initializez tabla si jocul
		Board b = new Board();

		while (true) {
			//Se citeste cate o linie pe rand
			line = sc.nextLine();

			// Ignoram comenzile specifice xboardului care nu afecteaza engine-ul
			if (line.contains("post"))
				continue;
			if (line.contains("hard"))
				continue;
			if (line.contains("easy"))
				continue;
			if (line.contains("time"))
				continue;
			if (line.contains("otim"))
				continue;
			if (line.contains("undo"))
				continue;

			// Dezactivam modul care trimite SIGINT
			if (line.contains("protover")) {
				System.out.println("feature sigint=0");
			}
			// Se verifica daca se primeste comanda new si se reface tabla la valorile intiale cu constructorul
			if (line.contains("new")) {
				b = new Board();
				mode = "black";
				continue;
			}

			// Se seteaza culoare cu care se joaca in functie de comanda primita si se citeste urmatoare linie
			if (line.contains("black")) {
				mode = "black";
				line = sc.nextLine();
			}
			if (line.contains("white")) {
				mode = "white";
				line = sc.nextLine();
			}
			// Daca se primeste comanda go se porneste engine-ul
			// Daca regele este amenintat se verifica daca poate fi mutat. Daca regele nu poate fi mutat se da resign
			// Altfel se verifica miscarile tuturor pieselor
			if (line.contains("go")) {
				if (b.verifyKing(mode)) {
					if (makeAKingMove(b, mode) == 0)
						System.out.println("resign");
					System.out.println("yes");
				} else
					makeAMove(b, mode);
				continue;
			} 
			// Se verifica daca linia citita este o mutare valida
			if (isMove(line, b) == 1) {
				b.move(line);
				// Dupa ce se realizeaza o miscare primita ca input, engine-ul realizeaza o mutare
				// Se verifica daca regele este amenintat. Daca regele nu poate fi mutat se da resign
				if (b.verifyKing(mode)) {
					System.out.println(mode + " in check");
					if (makeAKingMove(b, mode) == 0)
						System.out.println("resign");
				} else
					makeAMove(b, mode);
				continue;
			}
			// Daca se intra in modul force engine ul nu mai face mutari singur
			if (line.contains("force")) { 
				while (true) {
					line = sc.nextLine();
					// Se astepta "go", "new" sau un mod de joc(culoare) pentru a iesi din force mod
					if (line.contains("black") || line.contains("white")) { 													
						mode = line;
						break;
					}
					if (line.contains("go"))
						break;
					if (line.contains("new")) {
						b = new Board();
						mode = "black";
						break;
					}

					// Ignoram comenzile specifice xboardului care nu afecteaza engine-ul
					if (line.contains("post"))
						continue;
					if (line.contains("hard"))
						continue;
					if (line.contains("easy"))
						continue;
					if (isMove(line, b) == 1) {
						b.move(line);
						continue;
					}
				}
			}

			if (line.contains("quit"))
				break;
		}
		sc.close();
	}
}